src/engine/threaded_engine_perdevice.cc

```diff
@@ -264,9 +269,14 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
     this->is_worker_ = true;
     auto* task_queue = &(block->task_queue);
     RunContext run_ctx{ctx, nullptr};
+
     // execute task
     OprBlock* opr_block;
     ready_event->signal();
+
+    // Set default number of threads for OMP parallel regions initiated by this thread
+    OpenMP::Get()->on_start_worker_thread(true);
```

Comment:
Are you sure this would only affect this thread? Wouldn't you need to create a region explicitly?

Reply:
What do you mean by "Wouldn't you need to create a region explicitly"?  
We normally try to set num_threads, but mshadow just does the default.

The issue https://github.com/apache/incubator-mxnet/issues/9744 claims that set_omp_num_threads() is a per-thread value, used in parallel regions created by this thread only and not a global as I once though.  I checked openmp source code (in 3rdparty) and found this to be true.

Reply:
Note: set__nproc(thread, new_nth)

```cpp
#define set__nproc(xthread, xval)                                              \
  (((xthread)->th.th_current_task->td_icvs.nproc) = (xval))

/* Changes set_nproc */
void __kmp_set_num_threads(int new_nth, int gtid) {
  kmp_info_t *thread;
  kmp_root_t *root;

  KF_TRACE(10, ("__kmp_set_num_threads: new __kmp_nth = %d\n", new_nth));
  KMP_DEBUG_ASSERT(__kmp_init_serial);

  if (new_nth < 1)
    new_nth = 1;
  else if (new_nth > __kmp_max_nth)
    new_nth = __kmp_max_nth;

  KMP_COUNT_VALUE(OMP_set_numthreads, new_nth);
  thread = __kmp_threads[gtid];

  __kmp_save_internal_controls(thread);

  set__nproc(thread, new_nth);

.
.
.
```

Reply:
ok. then when num threads is changed it won't be propagated to all workers.
Will it ever be changed under current design?
for example when gpu is used for first time?

Reply:
Kernel::Launch listens to whether GPU has been used before, as well as other places that call into OpenMP::GetNum...
mshadow before and also now just does it’s own thing with the defaults. We actually reduce to vprocs/2 with CPUWorker call, so it’s less. we talked about whether to make mshadow comply but decided against it due to deprecation.
so except for mshadow, yes once gpu is used, most OMP operations drop to one CPU as before

Reply:
Made a unit test that verifies that this only affects the current thread and its "child" parallel regions